### PR TITLE
[PaxosStateLog] Wire up and Migrate To Verifying PSL

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Enable stage one of PSL migration, running in verification mode, still trusting the file system implementation as the source of truth.

**Implementation Description (bullets)**:
On startup does a one time migration from file to sqlite (migration will only be redone on future startups if discrepancy in greatest log entry is detected). Then, run in validation mode, where we do dual writing and dual reading to verify correctness of the new impl, but use old imps as the source of truth in case of mismatch.
 
**Testing (What was existing testing like?  What have you done to improve it?)**:
Added some integration tests that things are wired through correctly.

**Concerns (what feedback would you like?)**:
Do we need more tests?
This enables verification mode by default and without the option to not run it. Do we want to make it configurable?
Anything else I might have missed? I assume we don't want to migrate clients who use the leader block, but we could...

**Where should we start reviewing?**:
PaxosStateLogMigrationIntegrationTest.java & LocalPaxosComponents.java 

**Priority (whenever / two weeks / yesterday)**:
Today, I hope

### CODE REVIEW

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+        assertValueAbsent(nonMigratedRound, sqliteLog);
+        assertValuePresent(nonMigratedRound, fileBasedLog);
+        assertValueLearned(nonMigratedRound, learner);
+    }
+
+    @Test
+    public void legacyLogIsTheSourceOfTruthWhenValueAbsentFromLegacy() throws IOException {
+        int migratedRound = 100;
+        int rogueValue = 200;
```
Comment (jeremyk-91) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java:98]: 🎉

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        int otherRound = 100;
+        PaxosStateLog<PaxosValue> fileBasedLog = createFileSystemLog(CLIENT);
+        fileBasedLog.writeRound(round, valueForRound(round));
+
+        PaxosLearner learner = paxosComponents.learner(CLIENT);
+
+        Client otherClient = Client.of("other");
+        PaxosStateLog<PaxosValue> otherFileBasedLog = createFileSystemLog(otherClient);
+        otherFileBasedLog.writeRound(otherRound, valueForRound(otherRound));
+        fileBasedLog.writeRound(round, valueForRound(round));
```
Comment (jeremyk-91) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java:169]: I think it's already done in 162

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+    private void assertValueLearned(int secondRound, PaxosLearner learner) {
+        assertThat(learner.getLearnedValue(secondRound)).hasValue(valueForRound(secondRound));
+    }
+
+    private void assertValueNotLearned(int rogueValue, PaxosLearner learner) {
+        assertThat(learner.getLearnedValue(rogueValue)).isEmpty();
+    }
+
+    private void assertValuePresent(int rogueValue, PaxosStateLog<PaxosValue> sqliteLog) throws IOException {
```
Comment (jeremyk-91) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java:191]: Bit of a funky param name for something that is supposed to exist 😄

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                sqliteDirectory, UUID.randomUUID(), true);
+    }
+
+    private PaxosValue valueForRound(int i) {
+        return new PaxosValue("value", i, new byte[] {1});
+    }
+
+    private PaxosStateLog<PaxosValue> createFileSystemLog(Client client) {
+        Path dir = useCase.logDirectoryRelativeToDataDirectory(legacyDirectory).resolve(client.value());
+        String learnerLogDir = Paths.get(dir.toString(), PaxosTimeLockConstants.LEARNER_SUBDIRECTORY_PATH).toString();
```
Comment (jeremyk-91) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java:215]: nit: `dir.resolve(PaxosTimeLockConstants.LEARNER_SUBDIRECTORY_PATH).toString()`?

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        return new PaxosStateLogImpl<>(learnerLogDir);
+    }
+
+    private PaxosStateLog<PaxosValue> createSqliteLog(PaxosStorageParameters parameters) {
+        Supplier<Connection> conn = SqliteConnections
+                .createDefaultNamedSqliteDatabaseAtPath(parameters.sqliteBasedLogDirectory());
+        return SqlitePaxosStateLog.create(parameters.namespaceAndUseCase(), conn);
+    }
+
+    private PaxosStorageParameters getParametersForClient(Client client) {
```
Comment (jeremyk-91) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java:225]: consider `@VisibleForTesting` in other class

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/LocalPaxosComponents.java [hunk 6]
```diff
@@ -50,12 +52,14 @@
...
-            Path baseLogDirectory,
+            Path legacyLogDirectory,
+            Path sqliteLogDirectory,
             UUID leaderUuid,
             boolean canCreateNewClients) {
         this.metrics = metrics;
         this.paxosUseCase = paxosUseCase;
-        this.baseLogDirectory = baseLogDirectory;
+        this.baseLogDirectory = legacyLogDirectory;
+        this.sqliteLogDirectory = paxosUseCase.logDirectoryRelativeToDataDirectory(sqliteLogDirectory).toAbsolutePath();
```
Comment (jeremyk-91) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/LocalPaxosComponents.java:62]: This should just be `sqliteLogDirectory`: we don't need to have separate SQLite instances for each of the use cases.

File: leader-election-impl/src/test/java/com/palantir/paxos/PaxosConsensusTestUtils.java [hunk 7]
```diff
@@ -59,14 +62,15 @@ public static PaxosTestState setup(int numLeaders,
         for (int i = 0; i < numLeaders; i++) {
             failureToggles.add(new AtomicBoolean(false));
 
-            PaxosLearner learner = PaxosLearnerImpl.newLearner(getLearnerLogDir(i));
+            PaxosLearner learner = PaxosLearnerImpl
+                    .newVerifyingLearner(getLearnerStorageParameters(i), PaxosKnowledgeEventRecorder.NO_OP);
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/paxos/PaxosConsensusTestUtils.java:66]: 🎉

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java (hunk 4 above)
Comment (gmaretic) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosStateLogMigrationIntegrationTest.java:215]: I am pretty sure you are correct, but am not about to refactor and cause massive data corruption :P

File: leader-election-impl/src/test/java/com/palantir/paxos/PaxosStateLogMigratorTest.java [hunk 8]
```diff
@@ -93,19 +93,44 @@ public void migrationDeletesExistingState() {
...
+        List<PaxosValue> unExpectedValues = insertValuesWithinBounds(1, 2, source);
         migrateFrom(source);
 
+        assertThat(target.getLeastLogEntry()).isNotEqualTo(source.getLeastLogEntry());
+        expectedValues.forEach(value -> assertThat(getPaxosValue(target, value.seq)).isEqualTo(value));
+        unExpectedValues.forEach(value -> assertThat(readRoundUnchecked(target, value.seq)).isNull());
+    }
+
+    @Test
+    public void migirateAgainIfMismatchDetected() throws IOException {
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/paxos/PaxosStateLogMigratorTest.java:115]: nit

```suggestion
    public void migrateAgainIfMismatchDetected() throws IOException {
```

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLogMigrationState.java [hunk 9]
```diff
@@ -54,29 +56,45 @@ private void initialize() {
...
+    private Function<Queries, Boolean> migrateToState(States state) {
+        return dao -> {
+            assertCurrentStateAtMost(dao, state);
+            return dao.migrateToVersion(namespace, useCase, state.getSchemaVersion());
+        };
+    }
+
+    private void assertCurrentStateAtMost(Queries dao, States state) {
+        dao.getVersion(namespace, useCase).ifPresent(currentVersion ->
+                Preconditions.checkState(currentVersion <= state.getSchemaVersion()));
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLogMigrationState.java:95]: Could we add a bit more information (e.g. namespace, use-case, current/intended version) as args?

**Pull request merged**